src/drafts/SelectPanel2/SelectPanel.tsx

```diff
@@ -150,13 +155,32 @@ const Panel: React.FC<SelectPanelProps> = ({
 
   /* Dialog */
   const dialogRef = React.useRef<HTMLDialogElement>(null)
-  if (internalOpen) dialogRef.current?.showModal()
-  else dialogRef.current?.close()
+
+  // sync dialog open state (imperative) with internal component state
+  React.useEffect(() => {
+    if (internalOpen) dialogRef.current?.showModal()
+    else if (dialogRef.current?.open) dialogRef.current.close()
+  }, [internalOpen])
 
   // dialog handles Esc automatically, so we have to sync internal state
-  React.useEffect(() => dialogRef.current?.addEventListener('close', onInternalClose))
+  // but it doesn't call onCancel, so have another effect for that!
+  React.useEffect(() => {
+    const dialogEl = dialogRef.current
+    dialogEl?.addEventListener('close', onInternalClose)
+    return () => dialogEl?.removeEventListener('close', onInternalClose)
+  })
```

Comment:
Do we need these effects to run every render? 🤔

Reply:
Good question! I'll double check